CustomCollections/CustomCollections/CustomHashTable.cs

```diff
@@ -0,0 +1,220 @@
+                }
+                var position = GetPosition(key);
+                _hashTable[position] = new KeyValuePair<TKey, TValue>(key, value);
+            }
+        }
+
+        public void CopyTo(KeyValuePair<TKey, TValue>[] array, int arrayIndex)
+        {
+            if (array == null)
+            {
+                throw new ArgumentNullException("array");
+            }
+            if (arrayIndex < 0 || array.Length - arrayIndex < Count)
+            {
+                throw new ArgumentOutOfRangeException("arrayIndex");
+            }
+            for (int i = 0; i < Capacity; i++)
+            {
+                if (_hashTable[i].Key != null && !EqualityComparer<TKey>.Default.Equals(_hashTable[i].Key, default(TKey)))
+                {
+                    array[arrayIndex] = _hashTable[i];
+                    arrayIndex++;
+                }
+            }
+        }
+
+        public void Add(KeyValuePair<TKey, TValue> item)
+        {
+            if (item.Key == null)
+            {
+                throw new ArgumentNullException("key");
+            }
+            if (!HasAvailableMemory()) IncreaseSize();
+            if (ContainsKey(item.Key))
+            {
+                throw new ArgumentException("key");
+            }
+            var position = GetPosition(item.Key);
+            _hashTable[position] = item;
+            Count++;
+        }
+
+        public void Add(TKey key, TValue value)
+        {
+            Add(new KeyValuePair<TKey, TValue>(key, value));
+        }
+
+        public bool Remove(KeyValuePair<TKey, TValue> item)
+        {
+            return Remove(item.Key);
+        }
+
+        public bool Remove(TKey key)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException("key");
+            }
+            var position = GetPosition(key);
+            if (_hashTable[position].Key != null && !EqualityComparer<KeyValuePair<TKey, TValue>>.Default.Equals(_hashTable[position], new KeyValuePair<TKey, TValue>()))
+            {
+                _hashTable[position] = new KeyValuePair<TKey, TValue>();
+                Count--;
+                return true;
+            }
+            return false;
+        }
+
+        public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator()
+        {
+            foreach (var item in _hashTable)
+            {
+                if (!item.Equals(null) || !EqualityComparer<KeyValuePair<TKey, TValue>>.Default.Equals(item, new KeyValuePair<TKey, TValue>()))
+                {
+                    yield return item;
+                }
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        private int GetPosition(TKey key)
+        {
+            var hash = key.GetHashCode();
+            var capacity = Capacity;
+            if (hash >= Capacity) capacity = hash + 1;
+            var pos = Math.Abs(hash % capacity);
```

Comment:
This method is very weird. If hash is > Capacity (which would be true all the time), you position would always be hash % (hash + 1), which is hash, which would be out of range since it bigger then capacity.